plugins/web/opentelemetry-plugin-react-load/src/BaseOpenTelemetryComponent.ts

```diff
@@ -0,0 +1,541 @@
+    BaseOpenTelemetryComponent._tracer = api.trace.getTracer(
+      name,
+      version ? version : VERSION
+    );
+  }
+
+  /**
+   * Sets the logger for all components being instrumented
+   * @param logger
+   */
+  static setLogger(logger: api.Logger): void {
+    BaseOpenTelemetryComponent._logger = logger;
+  }
+
+  /**
+   * Creates a new span as a child of the current parent span.
+   * If parent span is undefined, just the child is created.
+   * @param react React component currently being instrumented
+   * @param name Name of span
+   */
+  private _createSpanWithParent(
+    react: React.Component,
+    name: string
+  ): api.Span | undefined {
+    return BaseOpenTelemetryComponent._tracer.startSpan(name, {
+      attributes: this._getAttributes(react),
+      parent: this._getCreateParentSpan(react)?.span,
+    });
+  }
+
+  /**
+   * Creates a new span
+   * @param react React component currently being instrumented
+   * @param name Name of span
+   */
+  private _createSpan(
+    react: React.Component,
+    name: string
+  ): api.Span | undefined {
+    return BaseOpenTelemetryComponent._tracer.startSpan(name, {
+      attributes: this._getAttributes(react),
+    });
+  }
+
+  /**
+   * Provides instrumentation for a function
+   * @param react React component currently instrumenting.
+   * @param spanName Name to set the span of the instrumented function to.
+   * @param original Original function currently being wrapped.
+   * @parentName Name to set parent span to on error.
+   */
+  private _instrumentFunction(
+    react: React.Component,
+    spanName: string,
+    original: any
+  ) {
+    const span = this._createSpanWithParent(react, spanName);
+    let res;
+    try {
+      res = original();
+    } catch (err) {
+      if (span) {
+        span.setAttribute(AttributeNames.REACT_ERROR, err.stack);
+        span.end();
+      }
+      this._endParentSpan(react);
+      throw err;
+    }
+
+    if (span) {
+      span.end();
+    }
+    return res;
+  }
+
+  /**
+   * Ends the current parent span.
+   * @param react React component parent span belongs to.
+   */
+  private _endParentSpan(react: React.Component) {
+    const parentSpanData = this._parentSpanMap.get(react);
+    if (parentSpanData) {
+      parentSpanData.span.updateName(parentSpanData.name);
+      parentSpanData.span.end();
+      this._parentSpanMap.delete(react);
+    }
+  }
+
+  /**
+   * Returns attributes object for spans
+   * @param react React component currently being instrumented
+   **/
+  private _getAttributes(react: React.Component) {
+    let state = '';
+    try {
+      state = JSON.stringify(react.state);
+    } catch {
+      state = '{"message": "state could not be turned into string"}';
+    }
+    return {
+      [GeneralAttribute.COMPONENT]: this.moduleName,
+      [AttributeNames.LOCATION_URL]: window.location.href,
+      [AttributeNames.REACT_NAME]: react.constructor.name,
+      [AttributeNames.REACT_STATE]: state,
```

Comment:
Is this a requirement? I think the state can become quite large in some cases and this may cause quite some overhead.

Reply:
I added in state as an attribute because I found it a good way to decipher where/when spans were coming from in an application, would it be better to not include this attribute or only conditionally include it based on length?

Reply:
A length heuristic is likely confusing for users. Maybe you could define a protected property on the base class like `getCustomSpanAttributes: undefined | (state: State) => Attributes`, which is called with state iff it is defined, and its returned attributes are attached to the span? Or a simpler property like `addStateAsSpanAttribute: boolean` might be easier... It depends what use-case you're actually trying to solve.

Reply:
If the use-case is to help you debug while building the module, I would say to strip it out before merge. If it is actually impossible to tell which component generated the span, then maybe there is some other attribute that could cover that use-case like the component's `constructor.name` property.

Reply:
I see, I'm trying to solve the user side use case. I was thinking that since it's common in React to have multiple state updates in a row (for example setting setting the state to Loading, then setting it to Loaded once a request returns), these would all come from the same component so just having a name (which is an attribute right now too) might still make it hard for users to decipher the traces from each other. 
Could you clarify about your suggestion I think I'm misunderstanding it a bit, is the suggestion to still allow long states to be added as attributes but not add the state label at all when state is undefined?

Reply:
@dyladan @thgao was this addressed already or there are still things that left ?

Reply:
I haven't addressed this change yet, I was hoping to understand it a bit more and get some clarification first!